xapian-applications/omega/handler_libabw.cc

```diff
@@ -0,0 +1,168 @@
+		     const char* field,
+		     size_t len,
+		     string& out)
+{
+    if (size_t(end - start) > len && memcmp(start, field, len) == 0) {
+	start += len;
+	while (start != end && isspace(*start)) start++;
+	if (start != end && (end[-1] != '\r' || --end != start)) {
+	    if (!out.empty())
+		out.push_back(' ');
+	    out.append(start, end - start);
+	}
+    }
+}
+
+static void
+parse_metadata(const char* data,
+	       size_t len,
+	       string& author,
+	       string& title,
+	       string& keywords)
+{
+    const char* p = data;
+    const char* end = p + len;
+
+    while (p != end) {
+	const char* start = p;
+	p = static_cast<const char*>(memchr(p, '\n', end - start));
+	const char* eol;
+	if (p)
+	    eol = p++;
+	else
+	    p = eol = end;
+	if ((end - start) > 5 && memcmp(start, "meta:", 5) == 0) {
+	    start += 5;
+	    switch (*start) {
+		case 'i': {
+		    if (author.empty())
+			PARSE_FIELD(start, eol, "initial-creator", author);
+		    break;
+		}
+		case 'k': {
+		    PARSE_FIELD(start, eol, "keyword", keywords);
+		    break;
+		}
+	    }
+	} else if ((end - start) > 3 && memcmp(start, "dc:", 3) == 0) {
+	    start += 3;
+	    switch (*start) {
+		case 'c': {
+		    if (!author.empty())
+			author.clear();
+		    PARSE_FIELD(start, eol, "creator", author);
+		    break;
+		}
+		case 's': {
+		    PARSE_FIELD(start, eol, "subject", keywords);
+		    break;
+		}
+		case 't': {
+		    PARSE_FIELD(start, eol, "title", title);
+		    break;
+		}
+	    }
+	} else if ((end - start) > 8 && memcmp(start, "dcterms:", 8) == 0) {
+	    start += 8;
+	    PARSE_FIELD(start, eol, "available", keywords);
+	}
+    }
+}
+
+static void
+clear_text(string& str, const char* text)
+{
+    if (text) {
+	for (int i = 0; text[i] != '\0'; ++i)
+	    if (!isspace(text[i]) || (i && !isspace(text[i - 1])))
+		str.push_back(text[i]);
+	str.push_back('\n');
+    }
+}
+
+bool
+extract(const string& filename,
+	const string& mimetype,
+	string& dump,
+	string& title,
+	string& keywords,
+	string& author,
+	string& pages,
+	string& error)
+{
+    try {
+	RVNGFileStream input(filename.c_str());
+
+	if (!libabw::AbiDocument::isFileFormatSupported(&input)) {
+	    error = "Libabw Error: The format is not supported";
+	    return false;
+	}
+
+	bool succeed = true;
+	RVNGString metadata_dump, content_dump;
+
+	RVNGTextTextGenerator metadata(metadata_dump, true);
+	if (libabw::AbiDocument::parse(&input, &metadata)) {
+	    const char* data = metadata_dump.cstr();
+	    size_t len = metadata_dump.size();
+	    parse_metadata(data, len, author, title, keywords);
+	    succeed = true;
+	} else {
+	    error = "Libabw Error: Fail to extract metadata";
+	}
+
+	RVNGTextTextGenerator content(content_dump, false);
+	if (libabw::AbiDocument::parse(&input, &content)) {
+	    clear_text(dump, content_dump.cstr());
+	    succeed = true;
+	} else {
+	    if (!error.empty())
+		error.push_back('\n');
+	    error += "Libabw Error: Fail to extract text";
+	}
+	return succeed;
```

Comment:
`succeed` is initialised to `true`, and then set to `true` conditionally in two places - it never gets set anywhere else so what's the point of it?

Reply:
Thanks for pointing this out. It was a mistake. The plan is to initialise `succeed` to `false` and set it to `true` if libabw extracts metadata/content from the file.
I have corrected this.

Reply:
The revised code doesn't entirely make sense either.

The value in `error` is only used if this function indicates there was an error by returning `false`.  If it returns `true` that means "it worked", so it doesn't make sense to also try to report an error.

I think unless we have evidence that failure here is actually OK it makes more sense to be strict and just set `error` and `return false` right away if any call to `libabw::AbiDocument::parse()` fails.

If that turns out to reject certain abiword documents which can be usefully opened with abiword or libreoffice then we can relax the checks appropriately in response to bug reports.

If we start off being lax here it will be very hard to tighten up later - users will be more forgiving of a new feature being overly strict than an existing feature changing in a way which stops things working for them which were fine before.

(This is different to the situation for e.g. OpenDocument where we were reusing the existing parsers and matching how they were being applied currently made sense.)

Reply:
Thank you for the insight. I have changed this.

* Removed the bool variable (succeed) altogether
* Returns false if it fails to extract either metadata or content